src/plugins/profiling/server/routes/search_flamechart.ts

```diff
@@ -172,61 +173,70 @@ async function queryFlameGraph(
   // profiling-stacktraces is configured with 16 shards
   const nQueries = 8;
   const stackTraces = new Map<StackTraceID, StackTrace>();
+  const stackFrameDocIDs = new Set<string>(); // Set of unique FrameIDs
+  const executableDocIDs = new Set<string>(); // Set of unique executable FileIDs.
+  const stackTraceIDs = [...stackTraceEvents.keys()];
+  const chunkSize = Math.floor(stackTraceEvents.size / nQueries);
+  const chunks = chunk(stackTraceIDs, chunkSize);
 
-  await logExecutionLatency(
+  const stackResponses = await logExecutionLatency(
     logger,
     'mget query for ' + stackTraceEvents.size + ' stacktraces',
     async () => {
-      const promises = new Array(nQueries);
-      const chunkSize = Math.floor(stackTraceEvents.size / nQueries);
-      const stackTraceIDs = [...stackTraceEvents.keys()];
-
-      for (let i = 0; i < nQueries; i++) {
-        const func = async () => {
-          const chunk = stackTraceIDs.slice(chunkSize * i, chunkSize * (i + 1));
-          const res = await client.mget({
+      return await Promise.all(
+        chunks.map((ids) => {
```

Comment:
Love this one ❤️

Reply:
Same here. There's fewer nested `async`s, which makes it easier to follow the logic.